metals/src/main/scala/scala/meta/internal/pantsbuild/BloopPants.scala

```diff
@@ -0,0 +1,373 @@
+                    C.Mixed,
+                    addLibraryToBootClasspath = true,
+                    addCompilerToClasspath = false,
+                    addExtraJarsToClasspath = false,
+                    manageBootClasspath = true,
+                    filterLibraryFromClasspath = true
+                  )
+                )
+              )
+            ),
+            java = Some(C.Java(Nil)),
+            sbt = None,
+            test = None,
+            platform = Some(C.Platform.Jvm(C.JvmConfig(javaHome, Nil), None)),
+            resolution = Some(C.Resolution(libraryDependencySources)),
+            resources = None
+          )
+      }.toSeq
+      val byName = bloopProjects.map(p => p.name -> p).toMap
+      val transitiveClasspath = mutable.Map.empty[String, List[Path]]
+      val isVisited = mutable.Set.empty[String]
+      def getTransitiveClasspath(name: String): List[Path] = {
+        if (isVisited(name)) {
+          transitiveClasspath.getOrElse(name, Nil)
+        } else {
+          isVisited += name
+          val result = transitiveClasspath.getOrElseUpdate(
+            name, {
+              val buf = mutable.Set.empty[Path]
+              buf ++= byName(name).classpath
+              byName(name).dependencies.foreach { dep =>
+                buf ++= getTransitiveClasspath(dep)
+              }
+              val children = cycles.children.getOrElse(name, Nil)
+              children.foreach { child =>
+                buf ++= getTransitiveClasspath(child)
+              }
+              buf.toList.sorted
+            }
+          )
+          result
+        }
+      }
+
+      val fullClasspathProjects = bloopProjects.map { p =>
+        val children = cycles.children.getOrElse(p.name, Nil)
+        val extraSources = children.flatMap(child => byName(child).sources)
+        val extraDependencies = children.iterator
+          .flatMap(child => byName(child).dependencies)
+          .filter(_ != p.name)
+          .toSeq
+        p.copy(
+          classpath = getTransitiveClasspath(p.name),
+          sources = (p.sources ++ extraSources).distinct,
+          dependencies = (p.dependencies ++ extraDependencies).distinct
+        )
+      }
+
+      var generatedProjects = Set.empty[Path]
+      fullClasspathProjects.foreach { bloop =>
+        if (!cycles.parents.contains(bloop.name)) {
+          val out = bloopRoot.resolve(makeFilename(bloop.name) + ".json")
+          val json = C.File(BloopVersion, bloop)
+          _root_.bloop.config.write(json, out)
+          generatedProjects += out
+        }
+      }
+      cleanOldBloopFiles(bloopRoot, generatedProjects)
+      Some(BloopInstallResult(fullClasspathProjects.size))
+    } else {
+      None
+    }
+  }
+
+  def cleanOldBloopFiles(
+      bloopRoot: Path,
+      generatedProjects: Set[Path]
+  ): Unit = {
+    val ls = Files.list(bloopRoot)
+    try {
+      ls.forEach { path =>
+        if (Files.isRegularFile(path) &&
```

Comment:
those conditions could be put in `filter` clauses for greater readability

Reply:
Done.